File: 2018/03/Main.hs
module Main where

import Data.List.Split (splitOn)
import Data.Map (elems, fromListWith)
import Data.Set (Set, size, (\\), union, unions, singleton, toList)
import Control.Arrow ((&&&))

main :: IO()
main = do
    contents <- readFile "input.txt"
    print $ (part1 &&& part2) $ parse contents

parse :: String -> [[Int]]
parse = map parseLine . lines

parseLine :: String -> [Int]
parseLine line = map read [tail i, x, init y, w, h]
    where
        [i, _, pos, dims] = splitOn " " line
        [x, y]            = splitOn "," pos
        [w, h]            = splitOn "x" dims

part1 :: [[Int]] -> Int
part1 = length . overlapping . fabric

part2 :: [[Int]] -> Int
part2 =
      head . toList                                               -- Change type from Set Int to Int
    . uncurry (\\) . (unions &&& (unions . overlapping)) . fabric -- Calculate answer

fabric :: [[Int]] -> [Set Int]
fabric =
      elems
    . fromListWith union
    . concatMap (\[i, x, y, w, h] -> [ ((x', y'), singleton i)
                                     | x' <- take w [x..],
                                       y' <- take h [y..]])

overlapping :: [Set Int] -> [Set Int]
overlapping = filter ((> 1) . size)

File: 2018/03/main.py
import re
from collections import defaultdict


def parse(text=open(0)):
    claims = defaultdict(set)

    for line in text:
        i, X, Y, W, H = parse_ints(line)
        for x in range(W):
            for y in range(H):
                claims[(X+x, Y+y)] |= {i}

    return claims


def parse_ints(line):
    return map(int, re.findall(r'\d+', line))


def overlapping(claims):
    return [c for c in claims.values() if len(c) >= 2]


def not_overlapping(claims):
    return set.union(*claims.values()) - set.union(*overlapping(claims))


claims = parse()

print(len(overlapping(claims)),
      [*not_overlapping(claims)][0])
